src/Chapter3.hs

```diff
@@ -480,6 +552,80 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle = Castle
+    { name :: String
+    } deriving (Show)
+
+data Wall = Wall deriving (Show)
+data Church = Church deriving (Show)
+data Library = Library deriving (Show)
+data People
+    = OnePerson
+    | TwoPersons
+    | ThreePersons
+    | FourPersons
+    deriving (Show)
+
+data House = House
+    { people :: People
+    } deriving (Show)
+
+data City
+    = CityWithCastleWallChurch
+        { castle  :: Castle
+        , walls    :: [Wall]
+        , church  :: Church
+        , houses  :: [House]
+        }
+    | CityWithCastleWallLibrary
+        { castle  :: Castle
+        , wall    :: [Wall]
+        , library :: Library
+        , houses  :: [House]
+        }
+    | CityWithChurch
+        { church  :: Church
+        , houses  :: [House]
+        }
+    | CityWithLibrary
+        { library :: Library
+        , houses  :: [House]
+        }
```

Comment:
Interesting approach of using a sum type for the city! 👍🏼 🧠 

I feel like it would be more convenient to work `City` if it was a record type instead of a sum type.

First, we need to decide with the castle. You have walls and castles separate, but that could lead to some problems.this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable

 Let's look at the task:

> Wall, but only if the city has a castle
> -- Optional castle with a name (as 'String')

That means that walls are very connected to the castle, and they could not exist separately. So we have three options for the castle:

1. Castle with a name
2. Castle with the name + walls
3. No castle

This looks like a sum type!:

```haskell
data Castle 
    = Castle String
    | WallCastle String
    | None
```

Note that your idea with castle suits here perfectly as well. But you need to somehow connect them together, for example another option could be:

```haskell
data CastleBuilding = Castle Building
    { castleName :: Castle
    , castleWall :: [Wall] 
    }
```

And in the city, we will have `Maybe CastleBuilding`.


Also, you see that only Church or only Library are allowed in the city. This is a good use case for a sum type, as it represents "OR" pattern very well:

```haskell
data ChurchOrLibrary
    = Church
    | Library
```

And you won't need to create additional types separately for each building, these constructors are enought.

So, in summary, instead of having two top-level constructors and duplicating all fields on both of them, you can define `City` as a record with optional castle:

After all that is done, Then let's build the city:

```haskell
data City = City
    -- Optional castle with a __name__ (as 'String')
    -- Wall, but only if the city has a castle
    { cityCastle :: Castle -- from the above
    -- OR you can do it with the other Castle type
    -- cityCastle :: Maybe Castle
    -- Church or library but not both
    , cityChurchOrLibrary :: ChurchOrLibrary
    -- any number of houses. Each house has one, two, three or four __people__ inside.
    , cityHouse s:: [House]
    }
```